### DISCUSSION TITLE
자바스크립트 비동기란?

### DISCUSSION THREAD

**Discussion by JUDONGHYEOK** [woowacourse-study/2022-woowahan-bansanghwe#132] [JavaScript]
## 비동기란
 비동기란 특정 코드를 수행하는 도중 다음 코드를 실행, 동기는 특정코드를 수행 완료한 후 다음 코드를 실행
- 비동기는 특정코드를 수행하는 도중 다음 코드를 실행
    - 콜백함수를 사용했기 때문이다.
    
    ```
    setTimeout(()=>{},100)
    ```
    
###  Callback: 되돌아와서 호출해라 라는 명령이다.

- 어느 시점에 되돌아와서 호출하나요?
- 제어권 역시 위임된다.
- 중첩된 콜백은 가독성이 떨어짐에 따라 디버깅도 힘들어진다.
- try Catch로 오류를 잡기 힘들다.

### Promise

- new Promise(executor) → resolve와 reject를 콜백으로 받는다.
- 콜벡핼
- catch로 에러잡기 가능
- promise가 callback을 대체하는 것은 아니고 콜백을 조금 더 편리하고 예측 가능하게 사용할 수 있도록 하는 객체이다.
- 비동기 코드를 비교적 순차적으로 보이게 작성 가능하다.
    - 콜백헬 대신 체이닝
- 비동기 흐름 제어를 호출하는 쪽에서 할 수 있다.
- 여전히 동기코드 보다는 가독성이 떨어진다.
- 에러핸들링의 문제
- resolve, reject 호출을 깜빡해서 프로미스가 결정되지 않는 문제는 해결 못함

### async/await(ES2017) - promise의 syntatic sugar

- promise를 yield하는 generator 함수
    - 이 promise가 generator의 iterator를 제어해서 비동기 작업이 완료되면 generator를 다시 끝까지 진행
        
        동기 코드처럼 읽히는 비동기 코드 작성
        
    - 일반적으로 사람은 동기적으로 생각하는 데 익숙하다
    - 에러 처리도 동기 코드에서처럼 가능
    - resolve, reject 호출 보장

### JS에서 비동기 이해가 필요해진 이유
- JS는 싱글 스레드 엔진 위에서 한번에 한 작업씩 동기적으로 실행되는 언어
- 그런데 이 언어를 사용하면서
    - 어떤 작업이 완료되기까지 시간이 걸리는 걸 기다리면서
        - 네트워크 요청
        - 파일로드
        - 타이머, 알람 등
        - 이벤트 처리 (이벤트 리스너)
    - JS의 하나뿐인 소중한 스레드를 block 시키지는 않고
    - 작업이 완료되면 그때 이후 필요한 작업을 진행할 수 있기 위해서는?
        
 자바스크립트는 기본적으로 싱글쓰레드이므로 동기적인 언어이다. 호이스팅이 된 이후부터 코드가 작성된 순서대로 하나하나 수행된다. 시간이 오래걸리는 작업들을 수행하려면 뒤에 나오는 작업들이 자연스럽게 블라킹된다. 따라서 이러한 문제 해결을 위해서 비동기가 등장했다. 시간이 걸리는 타이머나 API 후순위로 밀려도 되는 rAF등이 대표적인 비동기 작업이라고 불릴 수 있을 것 같다.

- 근데 rAF는 왜 비동기로 동작하는거지? → 한 프레임을 보장하기 위해서

[[이벤트 루프, 넌 누구냐](https://tecoble.techcourse.co.kr/post/2021-08-28-event-loop/)](https://tecoble.techcourse.co.kr/post/2021-08-28-event-loop/)

[[Event loop: microtasks and macrotasks](https://javascript.info/event-loop)](https://javascript.info/event-loop)

- 중간 과정에서는 async await가 굳이 필요없더라도 무방하다.
- 간단하게 then으로 처리가 가능한경우 없애도 된다.
    - 함수를 async로 만들 때는 곧 원래의 함수를 promise를 만드는 것이다.
    - 프로미스를 여러번 감싼다고 해서 마이크로 태스크큐가 감당하지 못할 정도는 아니다.
    - 하지만 협업에 문제가 발생할 수 있다.
- async/await를 사용하지 않아도 무방한 경우
    - promise를 그대로 리턴하는 중간 함수
    - 간단하게 then으로 처리하는 경우
    - 원래부터 promise를 반환하는 경우
- Promise.race() - 제일 먼저 도착하는 promise 만을 수행한다.
    - timeout을 정할 때 유용할 것 같다.
    
    ```tsx
    var p5 = new Promise(function(resolve, reject) {
        setTimeout(() => resolve('다섯'), 500);
    });
    var p6 = new Promise(function(resolve, reject) {
        setTimeout(() => reject(new Error('여섯')), 100);
    });
    
    Promise.race([p5, p6])
    .then(function(value) {
      // 실행되지 않음
    }, function(error) {
      console.log(error.message); // "여섯"
      // p6이 더 빠르므로 거부함
    });
    ```

**Comment by al-bur** [comment]
비동기란?

**Comment by al-bur** [comment]
왜 자바스크립트 개발자들에게 비동기가 중요할까?

**Comment by al-bur** [comment]
왜 rendering이 blocking이 될까요?

**Comment by al-bur** [comment]
어떠한 작업들이 브라우저 환경에서 비동기로 작동되나요?

**Comment by al-bur** [comment]
어떤 방식으로 비동기 처리 하시나요?

**Comment by al-bur** [comment]
promise.all, promise.race

**Reply by al-bur** [reply]
promise.all, promise.race을 사용해본적은?

**Comment by al-bur** [comment]
프로젝트에서 비동기 처리를 해줄 때 어려웠던 점은?

**Comment by al-bur** [comment]
promise는 callback의 어떠한 문제점을 개선하기 위해 나왔나

**Comment by al-bur** [comment]
왜 콜백함수는 try catch로 에러를 잡기 힘들까요?

**Reply by al-bur** [reply]
[키워드] 동기, 비동기

**Reply by greenblues1190** [reply]
try문이 길면 어떤 단점?

**Comment by al-bur** [comment]
병렬적이라는 뜻은?

**Reply by greenblues1190** [reply]
병렬성 vs 동시성
